### ISSUE TITLE
Custom 404 page - Cannot read properties of undefined (reading 'translations')

### ISSUE THREAD

**Issue by BuckyBuck135** [florian-lefebvre/astro-i18n-test#36]
Hi Florian,
Hope you're well.
I have a project using your package: https://github.com/CodeStitchOfficial/Advanced-Astro-v4-i18n

**1. The problem**
While on dev mode, navigating to any page throws this error:   
    
<img width="641" alt="image" src="https://github.com/user-attachments/assets/2a2a1a10-2132-4a5d-9655-b68a1e4dad11">

The line ` at 404 (eval at instantiateModule (file:///C:/Users/user/Documents/-=%20Web%20dev%20=-/CODE/-=%20Templates%20=-/Advanced-Astro-v4-i18n/node_modules/vite/dist/node/chunks/dep-BzOvws4Y.js:52861:24), <anonymous>:19:12)` made me look into the custom 404 page. 

Deleting this page gets rid of the error. Good, but I want a translated 404 page lol. Also note that this error does not happen on [the deployed site](https://advanced-astro-kit-i18n.netlify.app/fr/), where the 404 pages work perfectly.

**2. My setup**
Following [the docs](https://astro-i18n.netlify.app/recipes/translated-404-pages/), I've set up two 404 pages as I want fully static, under `src/pages/404.astro` and `src/pages/fr/404.astro`. I've also configured the rewrites for Netlify.

```js
//404.astro

---
import BaseLayout from "@layouts/BaseLayout.astro";
import { getLocalePath, t } from "i18n:astro";

const title = t("notFound.pageNotFound");
---

<BaseLayout title={title} description={title}>
    <h1>{title}</h1>
    <a href={getLocalePath("/")}>{t("notFound.linkText")}</a>
</BaseLayout>
```
I tried to narrow it down by deleting all references to `t()` or `getLocalePath()`, but I still get an error, with a different problem (reading 'data')
<img width="638" alt="image" src="https://github.com/user-attachments/assets/824ea85a-af9d-4033-b3ed-83810a302cb3">

(The repo currently shows one 404 page located in `routes/`, and I've tried the above setup in a branch)

```

Here's my Astro config:

```js
// astro.config.mjs
export default defineConfig({
  integrations: [
    i18n({
      defaultLocale: "en",
      locales: ["fr", "en"],
      client: {
        data: true,
      },
    })
  ]
});
``` 

I've tried setting all client keys to true (data, paths and translations), to no avail (I don't really understand what they do to be honest)

One thing I haven't tried is using SSR in hybrid mode.

While this error does not seem to impede the deployed site, it's a hassle to deal with in development. Do you have any idea where it comes from?

thank you for your time.

**Comment by BuckyBuck135** [comment]
After some more trying to figure it out, I've narrowed it down to the error originating somewhere from the BaseLayout.astro in the custom 404 pages.
Then, down to the use of `getLocale` to pass to the lang attribute in html

```html
---
// import some other stuff
import { getLocale} from "i18n:astro";

const locale = getLocale();
---
<html lang={locale}>
 // markup
 </html>
```

It looks like getLocale() can't be accessed. I've started using `Astro.currentLocale`, which now gets rid of all errors when navigating! The original issue is now solved, in a way.

However, the 404 pages do not work as expected anymore:
 * in dev mode: they return an error `Cannot read properties of undefined (reading 'translations')` looks familiar~ :)
 * in preview mode: redirect does not seem to be working anymore: 404 defaults to EN, which is the default locale.

**Comment by florian-lefebvre** [comment]
Thanks for opening an issue! Tbh I don't have time to debug this nor fix it, so feel free to use workarounds and/or submit a fix if you find one

**Comment by BuckyBuck135** [comment]
Hi Florian,
Thanks for your candor. If you don't mind, I'll keep this issue open as i continue the investigation, and maybe someone might be able to chime in.
Good luck with what you're working on!

**Comment by florian-lefebvre** [comment]
Sure let's keep it open! I'll label it correctly now

**Comment by BuckyBuck135** [comment]
Hi,
I've narrowed it down and created some repros.

Hi everyone,
Posting here because I know that the maintainer Florian is busy at the moment.

Now I believe that the issue boils down to: is a custom 404 route supposed to return errors in dev mode (while working fine in production)?

There's also some chatter on [this Starlight thread](https://discord.com/channels/830184174198718474/1070481941863878697/1286303516989853792) that could be related to this issue. To summarize, [middleware is intentionally not run in dev mode on Astro](https://github.com/withastro/astro/pull/11273)

In the minimal repros below, I have followed the installation steps laid out in [the integration's docs](https://astro-i18n.netlify.app/recipes/translated-404-pages/).

0. npm create astro@latest
1. npm run astro add @astrolicious/i18n
2. updated astro.config.mjs with the i18n config fron the doc
```js
import { defineConfig } from "astro/config"
import i18n from "@astrolicious/i18n"

export default defineConfig({
    integrations: [
        i18n({
            defaultLocale: "en",
            locales: ["en", "fr"],
            client: {
                data: true,
                // paths: true,
                // translations
            }
        })
    ]
})
```
3. Use client usage utilities in `Layout.astro`
`import I18nClient from "@astrolicious/i18n/components/I18nClient.astro"`
`<I18nClient />`
4. Placed `index.html` in `routes/`
5. Created `common.json` files in `locales/`
6. Created rewrites in `public/_redirects`
7. Deployed

Now, when running `npm run dev`, hitting a 404 route throws errors:
* in `404.astro`, if not using any of the i18n imports, returns this error: 
`[ERROR] Using `<I18nClient />` requires adding the `i18n` integration to your Astro config.`
*  in `404.astro`, using them returns this error: 
`[ERROR] Cannot read properties of undefined (reading 'data')`

Here are the repros:
 * [Stackblitz](https://stackblitz.com/edit/withastro-astro-ykuw9u?file=src%2Froutes%2Findex.astro)
 * [Github repo](https://github.com/BuckyBuck135/astrolicious-i18n-test)
* [Deployed site - the 404 route works as intended here](https://astrolicious-i18n-test.netlify.app/)

Thank you for your time!

**Comment by florian-lefebvre** [comment]
It should probably work after https://github.com/withastro/astro/pull/12034 is merged, and a patch released!

**Comment by BuckyBuck135** [comment]
Fingers crossed :)

**Comment by BuckyBuck135** [comment]
The issue is now fixed on my end with an update to Astro v4.15.9.
